File: examples/before_after_cleaning.rs
//! Prove that a cleaning step actually improved the data.
//!
//! Run with:
//!
//! ```bash
//! cargo run --example before_after_cleaning
//! ```
//!
//! The scenario: you wrote a cleaning script. Did it help? Profile the raw file,
//! persist that report, run the cleaning, profile the result, and diff the two.
//! Persisting the "before" report matters: weeks later you can still answer
//! "what did this data look like when we onboarded it?" without keeping the raw
//! file around.

use std::collections::BTreeMap;
use std::io::Write;
use std::path::Path;

use anyhow::Result;
use dataprof::{ColumnProfile, ProfileReport, Profiler};

const RAW: &str = "\
signup_id,email,plan,monthly_eur
S-1,ada@example.com,pro,29.00
S-2,,free,0.00
S-3,grace@example.com,PRO,29.00
S-4,,,
S-5,alan@example.com,free,0.00
S-5,alan@example.com,free,0.00
S-7,,enterprise,-99.00
S-8,edsger@example.com,pro,29.00
";

/// What our (imaginary) cleaning script does: drop rows with no email, drop the
/// duplicated signup, normalise `plan` casing, and reject the negative price.
const CLEAN: &str = "\
signup_id,email,plan,monthly_eur
S-1,ada@example.com,pro,29.00
S-3,grace@example.com,pro,29.00
S-5,alan@example.com,free,0.00
S-8,edsger@example.com,pro,29.00
";

fn profile(path: &Path) -> Result<ProfileReport> {
    Ok(Profiler::new()
        .positive_columns(vec!["monthly_eur".to_string()])
        .identifier_columns(vec!["signup_id".to_string()])
        .analyze_file(path)?)
}

fn null_percentage(col: &ColumnProfile) -> f64 {
    100.0 * col.null_count as f64 / col.total_count.max(1) as f64
}

fn by_name(report: &ProfileReport) -> BTreeMap<&str, &ColumnProfile> {
    report
        .column_profiles
        .iter()
        .map(|c| (c.name.as_str(), c))
        .collect()
}

fn main() -> Result<()> {
    let dir = tempfile::tempdir()?;

    let raw_path = dir.path().join("signups_raw.csv");
    let clean_path = dir.path().join("signups_clean.csv");
    write!(std::fs::File::create(&raw_path)?, "{RAW}")?;
    write!(std::fs::File::create(&clean_path)?, "{CLEAN}")?;

    let before = profile(&raw_path)?;

    // Persist the baseline. `ProfileReport` is serde-serialisable, so a report is
    // just JSON: commit it, ship it to object storage, or attach it to a PR.
    let baseline_path = dir.path().join("baseline.json");
    serde_json::to_writer_pretty(std::fs::File::create(&baseline_path)?, &before)?;

    let before: ProfileReport = serde_json::from_reader(std::fs::File::open(&baseline_path)?)?;
    let after = profile(&clean_path)?;

    println!("## Cleaning summary\n");
    println!(
        "| | rows | quality |\n|---|---:|---:|\n| before | {} | {:.1} |\n| after | {} | {:.1} |",
        before.execution.rows_processed,
        before.quality_score().unwrap_or(0.0),
        after.execution.rows_processed,
        after.quality_score().unwrap_or(0.0),
    );

    let dropped = before.execution.rows_processed - after.execution.rows_processed;
    let gained = after.quality_score().unwrap_or(0.0) - before.quality_score().unwrap_or(0.0);
    println!("\n{dropped} row(s) dropped, {gained:+.1} quality points\n");

    println!("## Per-column null rate\n");
    println!("| column | before | after | delta |");
    println!("|---|---:|---:|---:|");

    let (before_cols, after_cols) = (by_name(&before), by_name(&after));
    for (name, before_col) in &before_cols {
        let Some(after_col) = after_cols.get(name) else {
            println!(
                "| {name} | {:.0}% | *dropped* | |",
                null_percentage(before_col)
            );
            continue;
        };
        let (b, a) = (null_percentage(before_col), null_percentage(after_col));
        println!("| {name} | {b:.0}% | {a:.0}% | {:+.0}pp |", a - b);
    }

    // The headline numbers can improve while a specific defect survives. Check the
    // defects you set out to fix, not just the score.
    println!("\n## Did the defects go away?\n");
    let checks: [(&str, bool); 3] = [
        (
            "duplicate signup_id",
            after_cols
                .get("signup_id")
                .and_then(|c| c.unique_count.map(|d| d < c.total_count - c.null_count))
                .unwrap_or(false),
        ),
        (
            "negative monthly_eur",
            after
                .quality
                .as_ref()
                .and_then(|q| q.metrics.accuracy.as_ref())
                .is_some_and(|a| a.negative_values_in_positive > 0),
        ),
        (
            "missing email",
            after_cols.get("email").is_some_and(|c| c.null_count > 0),
        ),
    ];

    for (defect, still_present) in checks {
        println!(
            "  {} {defect}",
            if still_present {
                "STILL THERE"
            } else {
                "fixed      "
            }
        );
    }

    Ok(())
}

File: python/examples/before_after_cleaning.py
"""Prove that a cleaning step actually improved the data.

Run with:

    uv run python python/examples/before_after_cleaning.py

The scenario: you wrote a cleaning script. Did it help? Profile the raw file,
persist that report, run the cleaning, profile the result, and diff the two.
Persisting the "before" report matters: weeks later you can still answer "what
did this data look like when we onboarded it?" without keeping the raw file.
"""

from __future__ import annotations

import tempfile
from pathlib import Path

import dataprof as dp

RAW = """\
signup_id,email,plan,monthly_eur
S-1,ada@example.com,pro,29.00
S-2,,free,0.00
S-3,grace@example.com,PRO,29.00
S-4,,,
S-5,alan@example.com,free,0.00
S-5,alan@example.com,free,0.00
S-7,,enterprise,-99.00
S-8,edsger@example.com,pro,29.00
"""

# What our (imaginary) cleaning script does: drop rows with no email, drop the
# duplicated signup, normalise `plan` casing, and reject the negative price.
CLEAN = """\
signup_id,email,plan,monthly_eur
S-1,ada@example.com,pro,29.00
S-3,grace@example.com,pro,29.00
S-5,alan@example.com,free,0.00
S-8,edsger@example.com,pro,29.00
"""


def profile(path: Path) -> dp.ProfileReport:
    return dp.profile(
        str(path),
        positive_columns=["monthly_eur"],
        identifier_columns=["signup_id"],
    )


def main() -> None:
    with tempfile.TemporaryDirectory() as tmp:
        tmp_path = Path(tmp)
        raw_path = tmp_path / "signups_raw.csv"
        clean_path = tmp_path / "signups_clean.csv"
        raw_path.write_text(RAW, encoding="utf-8")
        clean_path.write_text(CLEAN, encoding="utf-8")

        # Persist the baseline before touching the data. A saved report is plain
        # JSON: commit it, ship it to object storage, or attach it to a PR.
        baseline = tmp_path / "baseline.json"
        profile(raw_path).save(str(baseline))

        before = dp.ProfileReport.load(str(baseline))
        after = profile(clean_path)

        before_score = before.quality_score
        after_score = after.quality_score
        assert before_score is not None and after_score is not None

        print("## Cleaning summary\n")
        print("| | rows | quality |")
        print("|---|---:|---:|")
        print(f"| before | {before.rows} | {before_score:.1f} |")
        print(f"| after | {after.rows} | {after_score:.1f} |")

        dropped = before.rows - after.rows
        gained = after_score - before_score
        print(f"\n{dropped} row(s) dropped, {gained:+.1f} quality points\n")

        # `compare()` does the diffing for you: quality, schema, and per-column
        # deltas. `a` is this report, `b` is the other one, and `abs` is `b - a`
        # (signed, despite the name).
        delta = before.compare(after)
        print("## What changed\n")
        print(f"quality_score: {delta['quality_score']['abs']:+.1f}")
        for dimension, change in sorted(delta["dimensions"].items()):
            if change["abs"]:
                print(f"  {dimension:<14} {change['abs']:+.1f}")

        if delta["schema"]["removed"]:
            print(f"\ncolumns dropped by cleaning: {delta['schema']['removed']}")

        print("\n## Per-column null rate\n")
        print("| column | before | after | delta |")
        print("|---|---:|---:|---:|")
        for name, change in sorted(delta["columns"].items()):
            print(
                f"| {name} | {change['null_pct_a']:.0f}% | {change['null_pct_b']:.0f}% "
                f"| {change['null_pct_delta']:+.0f}pp |"
            )

        # The headline numbers can improve while a specific defect survives. Check
        # the defects you set out to fix, not just the score.
        print("\n## Did the defects go away?\n")
        columns = after.column_profiles
        signup_id = columns["signup_id"]
        quality = after.quality
        assert quality is not None and quality.accuracy is not None
        checks = {
            "duplicate signup_id": (
                signup_id.unique_count is not None
                and signup_id.unique_count < signup_id.total_count - signup_id.null_count
            ),
            "negative monthly_eur": quality.accuracy["negative_values_in_positive"] > 0,
            "missing email": columns["email"].null_count > 0,
        }
        for defect, still_present in checks.items():
            print(f"  {'STILL THERE' if still_present else 'fixed      '} {defect}")


if __name__ == "__main__":
    main()
